### ISSUE TITLE
Measure where the pinned layer-streaming store is actually accounted (shmem vs anon) — the #622 mechanism is inferred, not verified

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#655]
The #622 fix rests on a reading of *where* the layer-streaming pinned host store is accounted, and that reading is an **inference that nobody has been able to test on hardware**. It is consistent with every number in the thread and it is not verified. This issue exists so it is either confirmed or overturned on a machine, rather than resting on a converging argument.

## What is established, and what is not

**Verified in code** (@umran666 on #622, greppable rather than arguable): the store is allocated exclusively via `torch.empty(..., device="cpu", pin_memory=True)` — the CUDA host allocator, `cudaHostAlloc`. There is no `shm_open`, no `multiprocessing.shared_memory`, and no tmpfs-backed file anywhere in the streaming path.

**Observed** (reporter's kernel OOM dumps on #622, torch 2.10, sm_120, kernel 7.0): `shmem-rss` 11.65-13.99 GB against `anon-rss` under 300 MB. The pinned store was accounted in the **shmem** class, not the anon class.

**Inferred, and this is the load-bearing part**: the NVIDIA driver backs large pinned host allocations with kernel-internal shmem objects — accounted as `Shmem`/`shmem-rss`, **not bound to any tmpfs mount** (which is why `df /dev/shm` stayed at 0% during the reporter's runs and why the mount's size could not have been the bound), and unreclaimable, hence the OOM killer rather than a graceful allocation failure.

If plain `cudaHostAlloc` pages were anonymous mlocked memory, the dumps would have shown the opposite split. So something has to explain the accounting; driver-internal shmem is the explanation that fits, and it has not been measured.

## Why it matters even though the shipped fix does not depend on it

The fix in #644 is deliberately **mechanism-agnostic**: store + resident extras against a fraction of total physical RAM, alongside the existing dynamic free-RAM check. That bound is correct under either accounting, which is why it shipped without this question being closed — and two independent implementations (#644, #646) converged on it.

What *does* depend on the inference is the decision **not** to add an `os.statvfs("/dev/shm")` check. The case against statvfs is that it measures a mount which is not the allocation channel, and it is wrong in both directions: a false positive on a large host with a small custom mount (forced to disk needlessly) and a false negative on a `--shm-size=64G` container on a small host. **If the experiment below comes back the other way, that decision reopens with evidence** — and I would rather that be findable in an open issue than buried in a closed PR.

## The experiment

Ten lines. Needs a **Linux host with CUDA** — nobody in the #622 thread has one (the reporter has WSL2, the analyst and I are on Windows).

```python
import torch
# BEFORE:
#   grep -E 'RssAnon|RssShmem' /proc/self/status
#   grep -E 'Shmem|Mlocked|Unevictable' /proc/meminfo
#   df /dev/shm
t = torch.empty(4 * 1024**3, dtype=torch.uint8, device="cpu", pin_memory=True)
t.fill_(1)   # fault every page in
# AFTER: read the same five numbers again
```

**Predicted under the current reading:**

- `RssShmem` climbs by roughly 4 GB
- `RssAnon` roughly unchanged
- `Unevictable` / `Mlocked` move little
- **`/dev/shm` usage unchanged**

**What would overturn it:** `/dev/shm` usage climbing by roughly 4 GB. That would mean the tmpfs mount really is the allocation channel on that stack, the mount's size really can bound the store, and a `statvfs` check belongs in the pre-flight after all.

Either result is a useful answer. A null result — numbers that match neither prediction — is also worth posting, because it would mean the accounting depends on something neither reading models.

## Acceptance criteria

- [ ] The five readings, before and after, from a Linux + CUDA host, with `torch`, driver and kernel versions stated.
- [ ] A statement of which prediction the numbers support, or that they support neither.
- [ ] If `/dev/shm` moves: reopen the statvfs question on #622 with this data, and say what the correct bound is.
- [ ] If it does not move: the inference becomes a measured fact and the `55%`-of-total-RAM ceiling's rationale can cite it instead of citing an argument.
- [ ] No code change is required by this issue. It is a measurement.

Worth adding: if you have such a box, please also report whether `stream_source: auto` on a store sized near 55% of total RAM behaves as the pre-flight predicts. That is a second, independent check on the shipped ceiling and costs one extra run.

Analysis and the experiment design are @umran666's, from [their comment on #622](https://github.com/MakazhanAlpamys/Soup/issues/622#issuecomment-5510190940). Filed by me so it does not disappear when #646 closes.

**Comment by tristangrech** [comment]
I'll take the measurement here. I have started a temporary Linux/CUDA pod on RunPod and will run the 4 GiB pinned-allocation probe, reading `/proc/self/status` inside the allocating process.

I'll record the PyTorch, CUDA, driver and kernel versions; process RSS, host memory counters, `/dev/shm` usage and container memory limits. I'll repeat the pinned run in a fresh process and add a pageable-memory control. Since this is a container, I'll distinguish process counters from host-wide counters that may include other tenants.

This covers the accounting experiment. The optional full streaming run near 55% of host RAM is outside this measurement.

**Comment by tristangrech** [comment]
The accounting prediction is reproduced on a Linux/CUDA RunPod container. Two fresh-process pinned runs each added exactly **4 GiB to `RssShmem`**, with no change to `RssAnon` or `/dev/shm` usage. The pageable control added exactly **4 GiB to `RssAnon`**, with no change to `RssShmem`.

Environment: NVIDIA L4; driver `570.195.03`; PyTorch `2.9.1+cu128`; CUDA runtime `12.8`; Python `3.12.3`; Ubuntu 24.04.3 container; Linux host kernel `6.14.0-33-generic`. Measurements taken on 2026-09-10 UTC. Neither `PYTORCH_ALLOC_CONF` nor `PYTORCH_CUDA_ALLOC_CONF` was set.

Each run used a fresh Python process. After CUDA initialization and a 1 MiB pinned warmup, it allocated `4 * 1024**3` bytes of `torch.uint8` host memory and touched every page with `fill_(1)`. `is_pinned()` was true for both pinned runs and false for the control. The process read its own `/proc/self/status` and `smaps_rollup` directly.

All numbers below are **KiB, before → after**. The allocation is 4,194,304 KiB.

| Counter | Pinned run 1 | Pinned run 2 | Pageable control |
|---|---:|---:|---:|
| Process `RssAnon` | 283648 → 283648 | 282624 → 282624 | 283648 → 4477952 |
| Process `RssShmem` | 10240 → 4204544 | 10240 → 4204544 | 10240 → 10240 |
| Process `VmLck` | 0 → 0 | 0 → 0 | 0 → 0 |
| Host `Shmem` | 381632 → 4575968 | 384296 → 4578760 | 384288 → 384456 |
| Host `Mlocked` | 27540 → 27540 | 27540 → 27572 | 27540 → 27540 |
| Host `Unevictable` | 30612 → 30612 | 30612 → 30644 | 30612 → 30612 |
| `/dev/shm` used | 0 → 0 | 0 → 0 | 0 → 0 |

`smaps_rollup.Pss_Shmem` independently increased by exactly 4 GiB in each pinned run and stayed unchanged for the control. Cgroup v2 `memory.stat.shmem` also increased by exactly 4 GiB for the pinned allocation. No cgroup OOM events occurred.

The container's `memory.max` was 93,999,996,928 bytes; its `/dev/shm` mount was 47,000,002,560 bytes. The script checked host and cgroup headroom before allocating. Host-wide counters can include activity from other tenants, so the small background changes in those columns should not be attributed to this allocation.

This supports the issue's predicted shared-memory accounting on this stack: the pinned allocation does not consume the `/dev/shm` mount. It does not establish the driver's internal implementation, prove the behavior on every driver/kernel, or measure reclaimability. I did not run the optional streaming test near 55% of host RAM or induce an OOM.

The temporary pod has been terminated. The exact probe below can be run twice with `--kind pinned` and once with `--kind pageable`; each invocation emits its environment and raw before/after counters as JSON. Probe SHA-256: `44bba1b76a9b41e927fe59666104bd44fcce63288f4af48887736118f7111c11`.


<details>
<summary>Exact probe used (save as probe.py)</summary>

```python
"""Soup #655: bounded pinned-host-memory accounting probe; one fresh process/run."""
import argparse
import gc
import json
import os
import platform
import subprocess
import time
from datetime import datetime, timezone
from pathlib import Path

import torch

GIB = 1024**3


def kb_fields(path, wanted):
    result = {}
    for line in Path(path).read_text().splitlines():
        key, _, value = line.partition(":")
        if key in wanted:
            result[key] = int(value.split()[0]) * 1024
    return result


def cgroup_snapshot():
    roots = [Path("/sys/fs/cgroup"), Path("/sys/fs/cgroup/memory")]
    for line in Path("/proc/self/cgroup").read_text().splitlines():
        _, controllers, relative = line.split(":", 2)
        if ".." in Path(relative).parts:
            continue
        if not controllers:
            roots.append(Path("/sys/fs/cgroup") / relative.lstrip("/"))
        elif "memory" in controllers.split(","):
            roots.append(Path("/sys/fs/cgroup/memory") / relative.lstrip("/"))
    results = {}
    for root in dict.fromkeys(roots):
        for filename in ("memory.current", "memory.max", "memory.limit_in_bytes",
                         "memory.usage_in_bytes", "memory.stat", "memory.events"):
            path = root / filename
            if path.is_file():
                results[str(path)] = path.read_text().strip()
    return results


def snapshot():
    shm = os.statvfs("/dev/shm")
    return {
        "utc": datetime.now(timezone.utc).isoformat(),
        "process_bytes": kb_fields("/proc/self/status", {
            "VmRSS", "VmLck", "RssAnon", "RssFile", "RssShmem"}),
        "smaps_rollup_bytes": kb_fields("/proc/self/smaps_rollup", {
            "Rss", "Pss", "Pss_Anon", "Pss_File", "Pss_Shmem", "Locked"}),
        "host_bytes": kb_fields("/proc/meminfo", {
            "MemTotal", "MemAvailable", "Shmem", "Mlocked", "Unevictable"}),
        "dev_shm_bytes": {
            "total": shm.f_blocks * shm.f_frsize,
            "used": (shm.f_blocks - shm.f_bfree) * shm.f_frsize,
            "available": shm.f_bavail * shm.f_frsize,
        },
        "cgroup": cgroup_snapshot(),
    }


def check_headroom(before, allocation_bytes):
    headrooms = [before["host_bytes"]["MemAvailable"]]
    cgroup = before["cgroup"]
    pairs = (("memory.max", "memory.current"),
             ("memory.limit_in_bytes", "memory.usage_in_bytes"))
    found = False
    for path, value in cgroup.items():
        for limit_name, usage_name in pairs:
            if Path(path).name != limit_name:
                continue
            usage = cgroup.get(str(Path(path).with_name(usage_name)))
            if usage is not None:
                found = True
                if value != "max":
                    headrooms.append(int(value) - int(usage))
    if not found:
        raise RuntimeError("Cannot verify container memory headroom")
    if min(headrooms) < 2 * allocation_bytes:
        raise RuntimeError("Less than 2x allocation headroom; refusing probe")
    return headrooms


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--kind", choices=("pinned", "pageable"), required=True)
    parser.add_argument("--gib", type=int, choices=(1, 4), default=4)
    args = parser.parse_args()
    torch.set_num_threads(1)
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is unavailable")
    environment = {
        "kernel": platform.release(), "platform": platform.platform(),
        "python": platform.python_version(), "torch": torch.__version__,
        "torch_cuda": torch.version.cuda, "gpu": torch.cuda.get_device_name(0),
        "nvidia_smi": subprocess.check_output([
            "nvidia-smi", "--query-gpu=name,driver_version,memory.total",
            "--format=csv,noheader"], text=True).strip(),
        "cgroup_membership": Path("/proc/self/cgroup").read_text().strip(),
        "allocator_environment": {key: os.environ.get(key) for key in (
            "PYTORCH_ALLOC_CONF", "PYTORCH_CUDA_ALLOC_CONF")},
        "os_release": Path("/etc/os-release").read_text().strip(),
    }
    torch.cuda.init()
    warmup = torch.empty(1024**2, dtype=torch.uint8, pin_memory=True)
    warmup.fill_(1)
    del warmup
    gc.collect()
    torch.cuda.synchronize()
    before = snapshot()
    allocation_bytes = args.gib * GIB
    headroom = check_headroom(before, allocation_bytes)
    start = time.monotonic()
    tensor = torch.empty(allocation_bytes, dtype=torch.uint8, device="cpu",
                         pin_memory=args.kind == "pinned")
    tensor.fill_(1)
    assert tensor.is_pinned() == (args.kind == "pinned")
    assert tensor[0].item() == 1 and tensor[-1].item() == 1
    time.sleep(0.25)
    after = snapshot()
    delta = {
        section: {key: after[section][key] - value
                  for key, value in before[section].items()}
        for section in ("process_bytes", "smaps_rollup_bytes", "host_bytes",
                        "dev_shm_bytes")
    }
    result = {
        "kind": args.kind, "allocation_bytes": allocation_bytes,
        "is_pinned": tensor.is_pinned(), "environment": environment,
        "headroom_bytes": headroom, "before": before, "after": after,
        "delta_bytes": delta, "duration_seconds": time.monotonic() - start,
        "note": "Process counters are local; host counters can include other tenants.",
    }
    del tensor
    gc.collect()
    print(json.dumps(result, indent=2), flush=True)


if __name__ == "__main__":
    main()
```

```bash
python probe.py --kind pinned --gib 4
python probe.py --kind pinned --gib 4
python probe.py --kind pageable --gib 4
```
</details>

**Comment by MakazhanAlpamys** [comment]
This is the answer the issue was opened for, and it lands on the prediction in every column. Thank you — the L4 pod was the thing nobody in the #622 thread could supply.

What the measurement settles, stated as narrowly as your data supports:

- **`RssShmem` +4 GiB exactly, `RssAnon` unchanged**, twice, in fresh processes. The pinned store is accounted in the shmem class.
- **The pageable control inverts it exactly** — `RssAnon` +4 GiB, `RssShmem` flat. That control is what turns this from a correlation into a discriminating result; without it, "shmem went up during a large allocation" would have been consistent with several readings.
- **`/dev/shm` usage unchanged at 0** while `Shmem` in `/proc/meminfo` climbed by 4 GiB. That is the specific claim the `statvfs` decision rested on: the tmpfs *mount* is not the allocation channel, so its size cannot bound the store.
- `smaps_rollup.Pss_Shmem` and cgroup v2 `memory.stat.shmem` each independently +4 GiB, `VmLck` flat throughout.

So the decision **not** to add an `os.statvfs("/dev/shm")` pre-flight check stands, and it now stands on a measurement rather than on a converging argument. The failure mode I was worried about — a false negative on `--shm-size=64G` — is confirmed to be the right worry: your container had a 47 GB `/dev/shm` mount and used **zero** of it while consuming 4 GiB of shmem accounting.

Two things I want on the record because they are what makes this reusable rather than a one-off:

1. You noted up front that host-wide counters in a container can include other tenants, and then leaned on the per-process and cgroup counters for the actual claim. That is why the small drift in the `Shmem` host row does not weaken anything.
2. You pinned the whole stack — L4, driver `570.195.03`, torch `2.9.1+cu128`, CUDA `12.8`, kernel `6.14.0-33` — which matters here more than usual, since the original observation came from torch 2.10 on `sm_120` / kernel 7.0 and the reading now holds across two quite different stacks.

**One follow-up, and it is yours if you want it.** `benchmarks/` is this project's public measurement record and is what the preprint points at; a result that overturns or confirms a shipped design decision belongs there rather than only in an issue thread. A small PR adding `benchmarks/gate-655-pinned-store-accounting.md` — your table verbatim, the probe script, the environment block, and one paragraph on what it settles about `statvfs` — plus the row in `benchmarks/README.md`, and I will merge it. Publish it as written, including the host-counter drift; the working record is the point, not a tidied-up report.

I will leave this open until that lands, then close it against the PR rather than against the comment, so the evidence is findable from the repository and not only from GitHub.

If you would rather not write the PR, say so and I will do it with the credit on your name — you did the expensive half.

**Comment by MakazhanAlpamys** [comment]
Closed by #787, contributed by @tristangrech, who ran the L4 measurement and has now published it as a record in `benchmarks/`: [`gate-655-pinned-store-accounting.md`](https://github.com/MakazhanAlpamys/Soup/blob/main/benchmarks/gate-655-pinned-store-accounting.md).

The settled question, stated as narrowly as the data supports: on that stack (L4, driver 570.195.03, torch 2.9.1+cu128, kernel 6.14), the pinned store is accounted in `RssShmem`, while a pageable control of the same size lands in `RssAnon`. So `statvfs` on `/dev/shm` is not a capacity bound for the pinned store there, and no `statvfs` preflight is being added. Other stacks and a full streaming run were not measured.
